src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/EnumExtensions.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal static class EnumExtensions
+{
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static unsafe void SetFlag<T>(ref this T value, T flag)
```

Comment:
There's a few code repetitions in this file, maybe the extracted logic can be extracted to something like this:

```cs
private static unsafe void ApplyFlagOperation<T>(ref T value, T flag, Action<byte*, byte*> operation)
            where T : unmanaged, Enum
        {
            var v = (T*)Unsafe.AsPointer(ref value);

            if (sizeof(T) == sizeof(byte))
            {
                operation((byte*)v, (byte*)&flag);
            }
            else if (sizeof(T) == sizeof(ushort))
            {
                operation((ushort*)v, (ushort*)&flag);
            }
            else if (sizeof(T) == sizeof(uint))
            {
                operation((uint*)v, (uint*)&flag);
            }
            else if (sizeof(T) == sizeof(ulong))
            {
                operation((ulong*)v, (ulong*)&flag);
            }
            else
            {
                Debug.Fail("Unexpected enum underlying type.");
            }
        }
```

Reply:
Actually, the duplicated logic is intentional. This is an `AggressiveInlining` method. So, it's expected that the JIT will perform the `sizeof` calculation at runtime and only inline the relevant branch. If the logic is shared, the performance gain will be lost, and the lambda would result in a delegate allocation. Also, the lambda that you suggested wouldn't work because `v` would need to be passed by reference.

I'll add comments explaining the nuance here.

Reply:
thanks I didn't realize that